src/webgpu/shader/execution/expression/call/builtin/textureStore.spec.ts

```diff
@@ -120,3 +120,387 @@ Parameters:
+function outOfBoundsValue(dim: GPUTextureDimension, type: string): string {
+  switch (dim) {
+    case '1d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords = -coords;
+        } else {
+          coords = coords + numTexels;
+        }`;
+      } else {
+        return `coords = coords + numTexels;`;
+      }
+      break;
+    }
+    case '2d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords.x = -coords.x;
+        } else {
+          coords.y = coords.y + height;
+        }`;
+      } else {
+        return `if gid.x % 3 == 1 {
+          coords.x = coords.x + width;
+        } else {
+          coords.y = coords.y + height;
+        }`;
+      }
+      break;
+    }
+    case '3d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords.x = -coords.x;
+        } else if gid.x % 5 == 0 {
+          coords.y = coords.y + height;
+        } else {
+          coords.z = coords.z + depth;
+        }`;
+      } else {
+        return `if gid.x % 3 == 1 {
+          coords.x = coords.x + width;
+        } else if gid.x % 5 == 1 {
+          coords.y = coords.y + height;
+        } else {
+          coords.z = 2 * depth;
+        }`;
+      }
+      break;
+    }
+  }
+  return ``;
+}
+
+function getMipTexels(numTexels: number, dim: GPUTextureDimension, mip: number): number {
+  let texels = numTexels;
+  if (mip === 0) {
+    return texels;
+  }
+  if (dim === '2d') {
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+  } else if (dim === '3d') {
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+  }
+  return texels;
+}
+
+const kDims: GPUTextureDimension[] = ['1d', '2d', '3d'];
+
+g.test('out_of_bounds')
+  .desc('Test that textureStore on out-of-bounds coordinates have no effect')
+  .params(u =>
+    u
+      .combine('dim', kDims)
+      .combine('coords', ['i32', 'u32'] as const)
+      .combine('mipCount', [1, 2, 3] as const)
+      .combine('mip', [0, 1, 2] as const)
+      .filter(t => {
+        if (t.dim === '1d') {
+          return t.mipCount === 1 && t.mip === 0;
+        }
+        if (t.dim === '3d') {
+          return t.mipCount <= 2 && t.mip < t.mipCount;
+        }
+        return t.mip < t.mipCount;
+      })
+  )
+  .fn(t => {
+    const texel_format = 'r32uint';
+    const num_texels = 4096;
+    const view_texels = getMipTexels(num_texels, t.params.dim, t.params.mip);
+
+    const texture_size = getTextureSize(num_texels, t.params.dim, 1);
+    const t_sizes: [number, number, number] = [
+      (texture_size as GPUExtent3DDict).width,
+      (texture_size as GPUExtent3DDict).height ?? 1,
+      (texture_size as GPUExtent3DDict).depthOrArrayLayers ?? 1,
+    ];
+    const mip_sizes = virtualMipSize(t.params.dim, t_sizes, t.params.mip);
+    const mip_size: GPUExtent3DDict = {
+      width: mip_sizes[0],
+      height: mip_sizes[1],
+      depthOrArrayLayers: mip_sizes[2],
+    };
```

Comment:
I don't know if you feel like doing this or not but, how about going and making `virtualMipSize` take a `GPUExtent3D`?

So this

```ts
/**
 * Compute the "virtual size" of a mip level of a texture (not accounting for texel block rounding).
 *
 * MAINTENANCE_TODO: Change output to Required<GPUExtent3DDict> for consistency.
 */
export function virtualMipSize(
  dimension: GPUTextureDimension,
  size: GPUExtent3D,
  mipLevel: number
): [number, number, number] {
  const temp = reifyExtent3D(size);
  const tempSize = [temp.width, temp.height, temp.depthOrArrayLayers];
  const shiftMinOne = (n: number) => Math.max(1, n >> mipLevel);
  switch (dimension) {
    case '1d':
      return [shiftMinOne(tempSize[0]), tempSize[1], tempSize[2]];
    case '2d':
      return [shiftMinOne(tempSize[0]), shiftMinOne(tempSize[1]), tempSize[2]];
    case '3d':
      return [shiftMinOne(tempSize[0]), shiftMinOne(tempSize[1]), shiftMinOne(tempSize[2])];
    default:
      unreachable();
  }
}
```

Then the code here code be this

```suggestion
    const texture_size = getTextureSize(num_texels, t.params.dim, 1);
    const mip_sizes = virtualMipSize(t.params.dim, texture_size, t.params.mip);
```


yea, the whole multiple ways to specify sizes is annoying. And we disagree on which is best :P   I prefer an array of 3 numbers (same as workgroup size) but others prefer an object with named properties.

IF you don't feel like changing it I'm happy to do it later.

Reply:
`virtualMipSize` got updated so the commit suggestion above *should work*

Reply:
Thanks for that PR. I've incorporated it as you suggested here.